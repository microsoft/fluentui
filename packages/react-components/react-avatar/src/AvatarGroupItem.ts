export * from './components/AvatarGroupItem/index';
