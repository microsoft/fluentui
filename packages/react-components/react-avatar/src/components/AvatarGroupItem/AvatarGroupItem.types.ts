import { AvatarGroupProps } from '../AvatarGroup/AvatarGroup.types';
import type { Avatar, AvatarSize } from '../../Avatar';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type AvatarGroupItemSlots = {
  root: NonNullable<Slot<'div', 'li'>>;

  /**
   * Avatar that represents a person or entity.
   */
  avatar: NonNullable<Slot<typeof Avatar>>;

  /**
   * Label used for the name of the AvatarGroupItem when rendered as an overflow item.
   * The content of the label, by default, is the `name` prop from the `avatar` slot.
   */
  overflowLabel: NonNullable<Slot<'span'>>;
};

/**
 * AvatarGroupItem Props
 */
export type AvatarGroupItemProps = Omit<ComponentProps<Partial<AvatarGroupItemSlots>, 'avatar'>, 'size' | 'shape'>;

/**
 * State used in rendering AvatarGroupItem
 */
export type AvatarGroupItemState = ComponentState<AvatarGroupItemSlots> & {
  /**
   * Whether the Avatar is an overflow item.
   *
   * @default false
   */
  isOverflowItem?: boolean;

  layout: AvatarGroupProps['layout'];
  size: AvatarSize;
};
