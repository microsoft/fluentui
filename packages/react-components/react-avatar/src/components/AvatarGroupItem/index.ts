export * from './AvatarGroupItem';
export * from './AvatarGroupItem.types';
export * from './renderAvatarGroupItem';
export * from './useAvatarGroupItem';
export * from './useAvatarGroupItemStyles.styles';
