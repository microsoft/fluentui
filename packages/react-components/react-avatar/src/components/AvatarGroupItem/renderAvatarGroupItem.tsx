import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { AvatarGroupItemState, AvatarGroupItemSlots } from './AvatarGroupItem.types';

/**
 * Render the final JSX of AvatarGroupItem
 */
export const renderAvatarGroupItem_unstable = (state: AvatarGroupItemState) => {
  const { slots, slotProps } = getSlots<AvatarGroupItemSlots>(state);

  return (
    <slots.root {...slotProps.root}>
      <slots.avatar {...slotProps.avatar} />
      {state.isOverflowItem && <slots.overflowLabel {...slotProps.overflowLabel} />}
    </slots.root>
  );
};
