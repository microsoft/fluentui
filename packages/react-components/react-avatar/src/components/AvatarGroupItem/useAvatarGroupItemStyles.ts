import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { tokens, typographyStyles } from '@fluentui/react-theme';
import { useSizeStyles } from '../../Avatar';
import { useFluent_unstable as useFluent } from '@fluentui/react-shared-contexts';
import type { AvatarGroupItemSlots, AvatarGroupItemState } from './AvatarGroupItem.types';
import type { AvatarGroupProps } from '../../AvatarGroup';
import type { AvatarSize } from '../../Avatar';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const avatarGroupItemClassNames: SlotClassNames<AvatarGroupItemSlots> = {
  root: 'fui-AvatarGroupItem',
  avatar: 'fui-AvatarGroupItem__avatar',
  overflowLabel: 'fui-AvatarGroupItem__overflowLabel',
};

const avatarGroupItemDividerWidthVar = '--fuiAvatarGroupItem__divider--width';

/**
 * Styles for the root slot
 */
const useRootStyles = makeStyles({
  base: {
    alignItems: 'center',
    display: 'inline-flex',
    flexShrink: 0,
    position: 'relative',
  },
  overflowItem: {
    ...shorthands.padding(tokens.spacingVerticalXS, tokens.spacingHorizontalXS),
  },
  nonOverflowItem: {
    ...shorthands.borderRadius(tokens.borderRadiusCircular),
  },
});

/**
 * Styles for the avatar slot
 */
const useAvatarStyles = makeStyles({
  nonOverflowItem: {
    position: 'absolute',
  },
  pie: {
    ...shorthands.borderRadius(0),
  },
});

/**
 * Styles for the label slot
 */
const useOverflowLabelStyles = makeStyles({
  base: {
    marginLeft: tokens.spacingHorizontalS,
    color: tokens.colorNeutralForeground1,
    ...typographyStyles.body1,
  },
});

/**
 * Styles for the stack layout
 */
const useStackStyles = makeStyles({
  thick: {
    boxShadow: `0 0 0 ${tokens.strokeWidthThick} ${tokens.colorNeutralBackground2}`,
  },
  thicker: {
    boxShadow: `0 0 0 ${tokens.strokeWidthThicker} ${tokens.colorNeutralBackground2}`,
  },
  thickest: {
    boxShadow: `0 0 0 ${tokens.strokeWidthThickest} ${tokens.colorNeutralBackground2}`,
  },
  xxs: { '&:not(:first-child)': { marginLeft: `calc(-1 * ${tokens.spacingHorizontalXXS})` } },
  xs: { '&:not(:first-child)': { marginLeft: `calc(-1 * ${tokens.spacingHorizontalXS})` } },
  s: { '&:not(:first-child)': { marginLeft: `calc(-1 * ${tokens.spacingHorizontalS})` } },
  l: { '&:not(:first-child)': { marginLeft: `calc(-1 * ${tokens.spacingHorizontalL})` } },
});

/**
 * Styles for the spread layout
 */
const useSpreadStyles = makeStyles({
  s: { '&:not(:first-child)': { marginLeft: tokens.spacingHorizontalS } },
  mNudge: { '&:not(:first-child)': { marginLeft: tokens.spacingHorizontalMNudge } },
  m: { '&:not(:first-child)': { marginLeft: tokens.spacingHorizontalM } },
  l: { '&:not(:first-child)': { marginLeft: tokens.spacingHorizontalL } },
  xl: { '&:not(:first-child)': { marginLeft: tokens.spacingHorizontalXL } },
});

/**
 * Styles for the pie layout
 */
const usePieStyles = makeStyles({
  base: {
    position: 'absolute',
  },
  slices: {
    // Two slices
    // 1st of 2 items
    '&:nth-of-type(1):nth-last-of-type(2)': {
      clipPath: `inset(0 calc(25% + (var(${avatarGroupItemDividerWidthVar}) / 2)) 0 25%)`,
      left: '-25%',
    },
    // 2nd of 2 items
    '&:nth-of-type(2):nth-last-of-type(1)': {
      clipPath: `inset(0 25% 0 calc(25% + (var(${avatarGroupItemDividerWidthVar}) / 2)))`,
      left: '25%',
    },

    // Three slices
    // 1st of 3 items
    '&:nth-of-type(1):nth-last-of-type(3)': {
      clipPath: `inset(0 calc(25% + (var(${avatarGroupItemDividerWidthVar}) / 2)) 0 25%)`,
      left: '-25%',
    },
    // 2nd of 3 items
    '&:nth-of-type(2):nth-last-of-type(2)': {
      // Since the two AvatarGroupItems on the right are scaled by 0.5, the divider width should not be halved.
      clipPath: `inset(0 0 var(${avatarGroupItemDividerWidthVar}) var(${avatarGroupItemDividerWidthVar}))`,
      left: '50%',
      transform: 'scale(0.5)',
      transformOrigin: '0 0',
    },
    // 3rd of 3 items
    '&:nth-of-type(3):nth-last-of-type(1)': {
      clipPath: `inset(var(${avatarGroupItemDividerWidthVar}) 0 0 var(${avatarGroupItemDividerWidthVar}))`,
      left: '50%',
      top: '50%',
      transform: 'scale(0.5)',
      transformOrigin: '0 0',
    },
  },
  rtlSlices: {
    // Two slices
    // 1st of 2 items
    '&:nth-of-type(1):nth-last-of-type(2)': {
      clipPath: `inset(0 25% 0 calc(25% + (var(${avatarGroupItemDividerWidthVar}) / 2)))`,
    },
    // 2nd of 2 items
    '&:nth-of-type(2):nth-last-of-type(1)': {
      clipPath: `inset(0 calc(25% + (var(${avatarGroupItemDividerWidthVar}) / 2)) 0 25%)`,
    },

    // Three slices
    // 1st of 3 items
    '&:nth-of-type(1):nth-last-of-type(3)': {
      clipPath: `inset(0 25% 0 calc(25% + (var(${avatarGroupItemDividerWidthVar}) / 2)))`,
    },
    // 2nd of 3 items
    '&:nth-of-type(2):nth-last-of-type(2)': {
      clipPath: `inset(0 var(${avatarGroupItemDividerWidthVar}) var(${avatarGroupItemDividerWidthVar}) 0)`,
      left: '0',
    },
    // 3rd of 3 items
    '&:nth-of-type(3):nth-last-of-type(1)': {
      clipPath: `inset(var(${avatarGroupItemDividerWidthVar}) var(${avatarGroupItemDividerWidthVar}) 0 0)`,
      left: '0',
    },
  },
  thick: { [avatarGroupItemDividerWidthVar]: tokens.strokeWidthThick },
  thicker: { [avatarGroupItemDividerWidthVar]: tokens.strokeWidthThicker },
  thickest: { [avatarGroupItemDividerWidthVar]: tokens.strokeWidthThickest },
});

/**
 * Apply styling to the AvatarGroupItem slots based on the state
 */
export const useAvatarGroupItemStyles_unstable = (state: AvatarGroupItemState): AvatarGroupItemState => {
  const { isOverflowItem, layout, size } = state;
  const { dir } = useFluent();

  const avatarStyles = useAvatarStyles();
  const overflowLabelStyles = useOverflowLabelStyles();
  const pieStyles = usePieStyles();
  const rootStyles = useRootStyles();
  const sizeStyles = useSizeStyles();

  const groupChildClassName = useGroupChildClassName(layout, size);

  const rootClasses = [rootStyles.base];

  if (!isOverflowItem) {
    rootClasses.push(rootStyles.nonOverflowItem);
    rootClasses.push(groupChildClassName);
    rootClasses.push(sizeStyles[size]);

    if (layout === 'pie') {
      rootClasses.push(pieStyles.base);

      if (size < 56) {
        rootClasses.push(pieStyles.thick);
      } else if (size < 72) {
        rootClasses.push(pieStyles.thicker);
      } else {
        rootClasses.push(pieStyles.thickest);
      }

      rootClasses.push(pieStyles.slices);

      if (dir === 'rtl') {
        rootClasses.push(pieStyles.rtlSlices);
      }
    }
  } else {
    rootClasses.push(rootStyles.overflowItem);
  }

  state.root.className = mergeClasses(avatarGroupItemClassNames.root, ...rootClasses, state.root.className);

  state.avatar.className = mergeClasses(
    avatarGroupItemClassNames.avatar,
    !isOverflowItem && avatarStyles.nonOverflowItem,
    layout === 'pie' && avatarStyles.pie,
    state.avatar.className,
  );

  if (state.overflowLabel) {
    state.overflowLabel.className = mergeClasses(
      avatarGroupItemClassNames.overflowLabel,
      overflowLabelStyles.base,
      state.overflowLabel.className,
    );
  }

  return state;
};

/**
 * Hook for getting the className for the children of AvatarGroup. This hook will provide the spacing and outlines
 * needed for each layout.
 */
export const useGroupChildClassName = (layout: AvatarGroupProps['layout'], size: AvatarSize): string => {
  const stackStyles = useStackStyles();
  const spreadStyles = useSpreadStyles();
  const layoutClasses = [];

  if (size) {
    if (layout === 'stack') {
      if (size < 56) {
        layoutClasses.push(stackStyles.thick);
      } else if (size < 72) {
        layoutClasses.push(stackStyles.thicker);
      } else {
        layoutClasses.push(stackStyles.thickest);
      }

      if (size < 24) {
        layoutClasses.push(stackStyles.xxs);
      } else if (size < 48) {
        layoutClasses.push(stackStyles.xs);
      } else if (size < 96) {
        layoutClasses.push(stackStyles.s);
      } else {
        layoutClasses.push(stackStyles.l);
      }
    } else if (layout === 'spread') {
      if (size < 20) {
        layoutClasses.push(spreadStyles.s);
      } else if (size < 32) {
        layoutClasses.push(spreadStyles.mNudge);
      } else if (size < 64) {
        layoutClasses.push(spreadStyles.l);
      } else {
        layoutClasses.push(spreadStyles.xl);
      }
    }
  }

  return mergeClasses(...layoutClasses);
};
