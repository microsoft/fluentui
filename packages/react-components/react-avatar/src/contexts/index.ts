export * from './AvatarGroupContext';
export * from './AvatarContext';
