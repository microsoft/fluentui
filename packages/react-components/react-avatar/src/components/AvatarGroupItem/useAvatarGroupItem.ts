import * as React from 'react';
import { Avatar } from '../Avatar/Avatar';
import { AvatarGroupContext, useAvatarGroupContext_unstable } from '../../contexts/AvatarGroupContext';
import { defaultAvatarGroupSize } from '../AvatarGroup/useAvatarGroup';
import { resolveShorthand } from '@fluentui/react-utilities';
import { useHasParentContext } from '@fluentui/react-context-selector';
import type { AvatarGroupItemProps, AvatarGroupItemState } from './AvatarGroupItem.types';

/**
 * Create the state required to render AvatarGroupItem.
 *
 * The returned state can be modified with hooks such as useAvatarGroupItemStyles_unstable,
 * before being passed to renderAvatarGroupItem_unstable.
 *
 * @param props - props from this instance of AvatarGroupItem
 * @param ref - reference to root HTMLElement of AvatarGroupItem
 */
export const useAvatarGroupItem_unstable = (
  props: AvatarGroupItemProps,
  ref: React.Ref<HTMLElement>,
): AvatarGroupItemState => {
  const groupIsOverflow = useAvatarGroupContext_unstable(ctx => ctx.isOverflow);
  const groupSize = useAvatarGroupContext_unstable(ctx => ctx.size);
  const layout = useAvatarGroupContext_unstable(ctx => ctx.layout);
  // Since the primary slot is not an intrinsic element, getPartitionedNativeProps cannot be used here.
  const { style, className, ...avatarSlotProps } = props;
  const size = groupSize ?? defaultAvatarGroupSize;
  const hasAvatarGroupContext = useHasParentContext(AvatarGroupContext);

  if (process.env.NODE_ENV !== 'production' && !hasAvatarGroupContext) {
    // eslint-disable-next-line no-console
    console.warn('AvatarGroupItem must only be used inside an AvatarGroup component.');
  }

  return {
    isOverflowItem: groupIsOverflow,
    layout,
    size,
    components: {
      root: groupIsOverflow ? 'li' : 'div',
      avatar: Avatar,
      overflowLabel: 'span',
    },
    root: resolveShorthand(props.root, {
      required: true,
      defaultProps: {
        style,
        className,
      },
    }),
    avatar: resolveShorthand(props.avatar, {
      required: true,
      defaultProps: {
        ref,
        size,
        color: 'colorful',
        ...avatarSlotProps,
      },
    }),
    overflowLabel: resolveShorthand(props.overflowLabel, {
      required: true,
      defaultProps: {
        // Avatar already has its aria-label set to the name, this will prevent the name to be read twice.
        'aria-hidden': true,
        children: props.name,
      },
    }),
  };
};
