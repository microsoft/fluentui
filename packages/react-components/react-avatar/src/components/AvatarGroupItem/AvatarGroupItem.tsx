import * as React from 'react';
import { renderAvatarGroupItem_unstable } from './renderAvatarGroupItem';
import { useAvatarGroupItem_unstable } from './useAvatarGroupItem';
import { useAvatarGroupItemStyles_unstable } from './useAvatarGroupItemStyles';
import type { AvatarGroupItemProps } from './AvatarGroupItem.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * The AvatarGroupItem component represents a single person or entity.
 * AvatarGroupItem should only be used in an AvatarGroup component.
 */
export const AvatarGroupItem: ForwardRefComponent<AvatarGroupItemProps> = React.forwardRef((props, ref) => {
  const state = useAvatarGroupItem_unstable(props, ref);

  useAvatarGroupItemStyles_unstable(state);
  return renderAvatarGroupItem_unstable(state);
});

AvatarGroupItem.displayName = 'AvatarGroupItem';
