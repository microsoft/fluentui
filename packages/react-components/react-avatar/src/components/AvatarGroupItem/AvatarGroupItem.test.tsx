import * as React from 'react';
import { render, screen } from '@testing-library/react';
import { AvatarGroupItem } from './AvatarGroupItem';
import { AvatarGroupContext } from '../../contexts';
import { isConformant } from '../../testing/isConformant';

const testId = 'testId';

const ContextWrapper: React.FC = ({ children }) => (
  <AvatarGroupContext.Provider value={{ isOverflow: true }}>{children}</AvatarGroupContext.Provider>
);

describe('AvatarGroupItem', () => {
  isConformant({
    Component: AvatarGroupItem,
    displayName: 'AvatarGroupItem',
    disabledTests: ['make-styles-overrides-win'],
    primarySlot: 'avatar',
    renderOptions: {
      wrapper: ContextWrapper,
    },
  });

  it('renders only an avatar when context provides false for isOverflow', () => {
    render(
      <AvatarGroupContext.Provider value={{ isOverflow: false }}>
        <AvatarGroupItem data-testid={testId} name="Katri Athokas" />
      </AvatarGroupContext.Provider>,
    );

    expect(screen.getByTestId(testId).textContent).toBe('KA');
  });

  it('renders a label and an avatar when context provides true for isOverflow', () => {
    render(
      <AvatarGroupContext.Provider value={{ isOverflow: true }}>
        <AvatarGroupItem name="Katri Athokas" overflowLabel={<span data-testid={testId}>Test Label</span>} />
      </AvatarGroupContext.Provider>,
    );

    expect(screen.getByTestId(testId).textContent).toBe('Test Label');
  });
});
