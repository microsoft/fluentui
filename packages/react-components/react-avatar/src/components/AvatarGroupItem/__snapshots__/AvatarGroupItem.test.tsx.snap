// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroupItem renders a default state 1`] = `
<div>
  <div
    class="fui-AvatarGroupItem"
  >
    Default AvatarGroupItem
  </div>
</div>
`;
